render/pixman/renderer.c

```diff
@@ -0,0 +1,339 @@
+#include "render/pixman.h"
+
+#include <wlr/render/interface.h>
+#include <wlr/util/log.h>
+
+#include <pixman.h>
+#include <wayland-server.h>
+#include <drm_fourcc.h>
+
+#include <stdlib.h>
+#include <assert.h>
+
+static const struct wlr_renderer_impl renderer_impl;
+
+static struct wlr_pixman_renderer *get_renderer(
+		struct wlr_renderer *wlr_renderer) {
+	assert(wlr_renderer->impl == &renderer_impl);
+	return (struct wlr_pixman_renderer *)wlr_renderer;
+}
+
+static struct wlr_pixman_buffer *get_buffer(
+		struct wlr_pixman_renderer *renderer, struct wlr_buffer *wlr_buffer) {
+	struct wlr_pixman_buffer *buffer;
+	wl_list_for_each(buffer, &renderer->buffers, link) {
+		if (buffer->buffer == wlr_buffer) {
+			return buffer;
+		}
+	}
+	return NULL;
+}
+
+static const struct wlr_texture_impl texture_impl;
+
+static struct wlr_pixman_texture *get_texture(
+		struct wlr_texture *wlr_texture) {
+	assert(wlr_texture->impl == &texture_impl);
+	return (struct wlr_pixman_texture *)wlr_texture;
+}
+
+static bool texture_is_opaque(struct wlr_texture *wlr_texture) {
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+
+	return texture->format == WL_SHM_FORMAT_XRGB8888;
+}
+
+static bool texture_write_pixels(struct wlr_texture *wlr_texture,
+		uint32_t stride, uint32_t width, uint32_t height,
+		uint32_t src_x, uint32_t src_y, uint32_t dst_x, uint32_t dst_y,
+		const void *data) {
+	assert(false && "todo texture_write_pixels");
+	return false;
+}
+
+static bool texture_to_dmabuf(struct wlr_texture *wlr_texture,
+		struct wlr_dmabuf_attributes *attribs) {
+	assert(false && "todo texture_to_dmabuf");
+	return false;
+}
+
+static void texture_destroy(struct wlr_texture *wlr_texture) {
+	if (wlr_texture == NULL) {
+		return;
+	}
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+
+	pixman_image_unref(texture->image);
+	free(texture);
+}
+
+static const struct wlr_texture_impl texture_impl = {
+	.is_opaque = texture_is_opaque,
+	.write_pixels = texture_write_pixels,
+	.to_dmabuf = texture_to_dmabuf,
+	.destroy = texture_destroy,
+};
+
+static uint32_t get_pixman_format_from_wl(enum wl_shm_format wl_fmt) {
+	uint32_t fmt = 0;
+	switch(wl_fmt) {
+	case WL_SHM_FORMAT_ARGB8888:
+	#if WLR_BIG_ENDIAN
+		fmt = PIXMAN_a8r8g8b8;
+	#elif WLR_LITTLE_ENDIAN
+		fmt = PIXMAN_b8g8r8a8;
+	#else
+	#error pixman endian error
+	#endif
+		break;
+	case WL_SHM_FORMAT_XRGB8888:
+	#if WLR_BIG_ENDIAN
+		fmt = PIXMAN_x8r8g8b8;
+	#elif WLR_LITTLE_ENDIAN
+		fmt = PIXMAN_b8g8r8x8;
+	#else
+	#error pixman endian error
+	#endif
+		break;
+	default:
+		// TODO: wl_shm_format to string?
```

Comment:
It's probably enough to just print the hex value, e.g.`"Unsupported wl_shm_format 0x%"PRIX32`. We don't really want to have a big table of `wl_shm_format` to strings, and could be missing entries in it anyways.